File: #0904.fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        
        map<int,int> m;
        int index = 0;
        int size = fruits.size();
        int result = 0;
        for (int i = 0; i < size; i++) {            
            if (m.size() <= 2) { 
                m[fruits[i]]++;
            }
            
            while (m.size() > 2 && index < size) {
                m[fruits[index]]--;
                if (!m[fruits[index]]) {
                    m.erase(fruits[index]);
                }
                index++;
            }
            result = std::max(result, i - index + 1);
        }
        return result;
    }
};

File: #0904.fruit-into-baskets.ts
function totalFruit(fruits: number[]): number {
    const n = fruits.length;
    const fruitCount = new Map<number, number>();
    let left = 0;
    let maxFruits = 0;
    for (let right = 0; right < n; right++) {
        fruitCount.set(fruits[right], (fruitCount.get(fruits[right]) || 0) + 1);
        while (fruitCount.size > 2) {
            const leftFruit = fruits[left];
            fruitCount.set(leftFruit, fruitCount.get(leftFruit)! - 1);
            if (fruitCount.get(leftFruit) === 0) {
                fruitCount.delete(leftFruit);
            }
            left++;
        }

        maxFruits = Math.max(maxFruits, right - left + 1);
    }
    return maxFruits;
};
